torch/_decomp/decompositions_for_rng.py

```diff
@@ -0,0 +1,270 @@
+from collections import defaultdict
+from typing import Callable, Dict
+
+import torch
+import torch._decomp as decomp
+from torch._ops import OpOverload
+from torch._prims_common import make_contiguous_strides_for
+
+
+aten = torch.ops.aten
+
+rng_decompositions: Dict[str, Dict[OpOverload, Callable]] = defaultdict(dict)
+
+
+def throw_if_philox_offset_did_not_advance(fn):
+    # Raises error if one forgot to advance the offset while adding a new decomposition.
+    def inner(*args, **kwargs):
+        old_relative_offset = PhiloxStateTracker.get_current_relative_offset()
+        out = fn(*args, **kwargs)
+        new_relative_offset = PhiloxStateTracker.get_current_relative_offset()
+        if old_relative_offset == new_relative_offset:
+            raise ValueError("Philox seed offset did not advance after decomposition")
+        return out
+
+    return inner
+
+
+def register_rng_decomposition(aten_op):
+    def inner(fn):
+        return decomp.register_decomposition(aten_op, rng_decompositions)(
+            throw_if_philox_offset_did_not_advance(fn)
+        )
+
+    return inner
+
+
+def throw_on_non_cuda(device):
+    raise RuntimeError(
+        f"You are trying to functionalize a {device.type} RNG operator but {device.type} does not "
+        f"use Philox/counter-based PRNG. Therefore, functionalizing a {device.type} RNG operator is "
+        "not supported. We are discussing the possibility of a Philox-based RNG implementation for CPU."
+    )
+
+
+def rand_offset_calculator(shape):
+    # For impl, look at the function calc_execution_policy in the file
+    # aten/src/ATen/native/cuda/DistributionTemplates.h. The impl was copied at
+    # commit hash 72aa0667bd16707d50eb8fa337092a1f5d11dfb6
+    numel = 1
+    for dim_size in shape:
+        numel *= dim_size
+
+    block_size = 256
+    unroll = 4
+    curand4_engine_calls = 4
+    device_property = torch.cuda.get_device_properties(torch.cuda.current_device())
+    blocks_per_sm = int(device_property.max_threads_per_multi_processor / block_size)
+    grid_size = int((numel + block_size - 1) / block_size)
+    grid_size = min(grid_size, device_property.multi_processor_count * blocks_per_sm)
+    offset = (
+        int((numel - 1) / (block_size * grid_size * unroll) + 1) * curand4_engine_calls
+    )
+    return offset
+
+
+# TODO - We have to register many more distributions here, and also higher level
+# ops like dropout which have fused implementation and can hide the rand inside.
+@register_rng_decomposition(aten.rand)
+def rand(shape, dtype=None, layout=torch.strided, device=None, pin_memory=False):
+    device = device or "cpu"
```

Comment:
```suggestion
    device = device or torchl.device("cpu")
```

Reply:
nit: `torch.device("cpu")` ?

Reply:
Should it be default device, not "cpu"? If e.g. we are in device context manager